### PR TITLE
chore: replace AddressBook with Roster in SignedState

### LINKED ISSUE #16906: Replace `AddressBook` with `Roster` in `SignedState`
**Background**
Replace all usages of `AddressBook` with `Roster` in `SignedState.java` 
* create a `SignedState.getRoster()` method that uses the `RosterRetriever` to get the latest active roster. 
* replace usages of `SignedState.getAddressBook()` with `SignedState.getRoster()`
  * This may require translating back to an addressbook in the calling class if the roster data structure is insufficient. 

**Acceptance Criteria**

* This refactor should be able to eliminate the java import of the `AddressBook` class. 
* Reconnect tests work.

### PR DESCRIPTION
**Pull request by timfn-hg**
**Description**:
Removes the usage of address books in SignedState and replaces it with the new roster.

**Related issue(s)**:

Fixes #16906 

**Notes for reviewer**:

**Checklist**

- [X] Documented (Code comments, README, etc.)
- [X] Tested (unit, integration, etc.)

### CODE REVIEW

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/roster/RosterRetriever.java
Comment (anthony-swirldslabs) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/roster/RosterRetriever.java:253]: I'm unsure if this is a good idea.

Generally, mutable collections are dangerous and should be avoided, especially in production code. And the RosterRetriever is a production code for the most part.

Specifically to the Roster, this data structure is considered immutable. Trying to create "mutable" instances leads to testing something that we've never designed or intended. And certainly, something that we (should) never use in production code. So from this perspective, the value of such tests is unclear.

I suggest to find a way to avoid creating mutable instances of this class.

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/signed/SignedState.java
Comment (anthony-swirldslabs) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/signed/SignedState.java:274]: I know that the `retrieveActiveOrGenesisRoster` may return null. However, previously this method guaranteed that the return value is non-null. I believe we may have code that relies on this. And even if we don't, I believe a `SignedState` should always have a non-null Roster - otherwise it couldn't have been signed in the first place because signatures are supposed to come from the nodes of the network, and the roster specifies the nodes. So if there's no nodes, there cannot be signatures, and hence the state cannot be signed. In other words, while `retrieveActiveOrGenesisRoster` could return null in theory, it should never return `null` when called on a (real, or properly mocked) SignedState.

Could we restore the `@NonNull` property of this method?

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/signed/SignedState.java
Comment (anthony-swirldslabs) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/signed/SignedState.java:730]: Minor: please note that the old code would only update the `signingWeight` instance variable if this code never threw any exceptions. Not sure if there's any code that relies on this fact, but your version changes this invariant. I'm totally fine with keeping your version as it avoids an extra long value on the stack, but ideally it would be great to confirm if this is safe to do.

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/snapshot/SignedStateFileWriter.java
Comment (anthony-swirldslabs) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/snapshot/SignedStateFileWriter.java:165]: Minor: we should probably file a ticket to revisit this part in the future because this method likely writes something in the form of a `config.txt` which is currently being deprecated. Ideally, we'd want to write a JSON of the Roster or something like that. But this may require evaluating current usages of this file and perhaps coordinating with DevOps/NodeOps/test frameworks that may use it.

File: platform-sdk/swirlds-platform-core/src/test/java/com/swirlds/platform/state/StateSigningTests.java
Comment (anthony-swirldslabs) [inline platform-sdk/swirlds-platform-core/src/test/java/com/swirlds/platform/state/StateSigningTests.java:94]: Minor: since the SignedState is now converted to use Rosters, do we want to update its generator to also only ever use rosters? This can be done in a separate PR if you like, in which case please file a new ticket to track this work.

File: platform-sdk/swirlds-platform-core/src/test/java/com/swirlds/platform/state/StateSigningTests.java
Comment (anthony-swirldslabs) [inline platform-sdk/swirlds-platform-core/src/test/java/com/swirlds/platform/state/StateSigningTests.java:332]: We no longer support this operation in the Roster world. It's immutable by definition. The only way to remove a node would be to create a new roster with an updated roster entries where that node is missing. This would yield a brand new instance of the Roster object. But we cannot and shouldn't do that in place in an already created Roster object.

File: platform-sdk/swirlds-platform-core/src/test/java/com/swirlds/platform/state/StateSigningTests.java
Comment (anthony-swirldslabs) [inline platform-sdk/swirlds-platform-core/src/test/java/com/swirlds/platform/state/StateSigningTests.java:474]: Same as above: Roster is an immutable object by definition/specification. So we can no longer update it in place. We can only create a new instance with whatever updates we want to apply.

This may invalidate certain tests and use case. Or one needs to be creative to see how to re-implement them w/o losing the immutability property of the Roster.

File: platform-sdk/swirlds-platform-core/src/testFixtures/java/com/swirlds/platform/test/fixtures/addressbook/RandomRosterBuilder.java
Comment (anthony-swirldslabs) [inline platform-sdk/swirlds-platform-core/src/testFixtures/java/com/swirlds/platform/test/fixtures/addressbook/RandomRosterBuilder.java:153]: Same as above. We shouldn't allow the mutability by the specification of this object.

File: platform-sdk/swirlds-platform-core/src/testFixtures/java/com/swirlds/platform/test/fixtures/addressbook/RosterTestUtils.java
Comment (anthony-swirldslabs) [inline platform-sdk/swirlds-platform-core/src/testFixtures/java/com/swirlds/platform/test/fixtures/addressbook/RosterTestUtils.java:107]: Please find a way to re-implement the tests w/o using this functionality. Or if that's totally impossible, then we could discuss this somewhere (the platform slack channel perhaps?) and maybe nuke the tests that are no longer relevant/supported.

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/roster/RosterRetriever.java
Comment (timfn-hg) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/roster/RosterRetriever.java:253]: I generally agree. The reason for making these mutable was to mimic how the address book supported adding and removing addresses from the same instance - and this was used a fair bit in testing. If the intention is to never mutate the roster, then I can remove this and re-write the tests to not do this either.

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/signed/SignedState.java
Comment (timfn-hg) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/signed/SignedState.java:730]: Are you referring to the old code in this method? If so, this is all new code that I added in a previous PR. I added the local `long` variable and then did the update to `signingWeight` in one operation. This was _not_ the same as the previous `pruneInvalidSignatures(AddressBook)`, but in this PR I did "revert" my change to mirror what was done in the original prune method that used the address book.

I can absolutely revert this back to my original code (single update), but it wouldn't match the original prune method. I made this change because I didn't know if anything was dependent on the original way of doing the recalculation. :)

File: platform-sdk/swirlds-platform-core/src/test/java/com/swirlds/platform/state/StateSigningTests.java
Comment (timfn-hg) [reply platform-sdk/swirlds-platform-core/src/test/java/com/swirlds/platform/state/StateSigningTests.java:332]: Like I mentioned above, this was all done to support what the original address book-based tests were doing. If the intent - which your comment makes it seem so - to not support such functionality for rosters, then I will remove that functionality and rework the tests.

File: platform-sdk/swirlds-platform-core/src/test/java/com/swirlds/platform/state/StateSigningTests.java
Comment (timfn-hg) [reply platform-sdk/swirlds-platform-core/src/test/java/com/swirlds/platform/state/StateSigningTests.java:94]: I'll create a new ticket. Updating `RandomSignedStateGenerator` was sorta blocked by your work and will become unblocked once your [PR](https://github.com/hashgraph/hedera-services/pull/17047) is merged in.

Comment (timfn-hg) [reply platform-sdk/swirlds-platform-core/src/test/java/com/swirlds/platform/state/StateSigningTests.java:94]: Created https://github.com/hashgraph/hedera-services/issues/17087

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/snapshot/SignedStateFileWriter.java
Comment (timfn-hg) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/snapshot/SignedStateFileWriter.java:165]: Created https://github.com/hashgraph/hedera-services/issues/17089

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/roster/RosterRetriever.java
Comment (edward-swirldslabs) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/roster/RosterRetriever.java:253]: Rosters must be immutable once built since their hash value is used as a reference to them.

Comment (timfn-hg) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/roster/RosterRetriever.java]: Got it! I've updated the PR to remove the mutable parts. Please re-review when you two get the chance.

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/signed/SignedState.java
Comment (anthony-swirldslabs) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/signed/SignedState.java:729]: Ah, I see. Thanks for clarifying. This indicates that this is safe then. I'm good with the current version of the code.

**Review by anthony-swirldslabs — DISMISSED** [review]

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/signed/SignedState.java
Comment (timfn-hg) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/signed/SignedState.java:267]: Done

**Review by anthony-swirldslabs — DISMISSED** [review]

**Comment by timfn-hg** [comment]
@anthony-swirldslabs : Can you please review one last time (hopefully). I had some merge conflicts after you merged your last set of changes and I want to make sure I didn't miss anything when I resolved the conflicts.

**Review by anthony-swirldslabs — APPROVED** [review]

**Pull request merged**
